knora/dsplib/utils/xml_upload.py

```diff
@@ -910,94 +918,340 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
 
-        res_iri_lookup[resource.id] = resclass_instance.iri
-        print(f"Created resource '{resclass_instance.label}' ({resource.id}) with IRI '{resclass_instance.iri}'")
+    return id2iri_mapping, failed_uploads
 
-    # update the resources with the stashed XML texts
-    if len(stashed_xml_texts) > 0:
-        print('Update the stashed XML texts...')
-    for resource, link_props in stashed_xml_texts.items():
+
+def update_stashed_xml_texts(
+    verbose: bool,
+    id2iri_mapping: dict[str, str],
+    con: Connection,
+    stashed_xml_texts: dict[XMLResource, dict[XMLProperty, dict[str, KnoraStandoffXml]]]
+) -> dict[XMLResource, dict[XMLProperty, dict[str, KnoraStandoffXml]]]:
+    """
+    After all resources are uploaded, the stashed xml texts must be applied to their resources in knora.
+
+    Args:
+        verbose: bool
+        id2iri_mapping: mapping of ids from the XML file to IRIs in knora
+        con: connection to knora
+        stashed_xml_texts: all xml texts that have been stashed
+
+    Returns:
+        nonapplied_xml_texts: the xml texts that could not be updated
+    """
+
+    print('Update the stashed XML texts...')
+    for resource, link_props in stashed_xml_texts.copy().items():
         print(f'Update XML text(s) of resource "{resource.id}"...')
-        res_iri = res_iri_lookup[resource.id]
-        try:
-            update_xml_texts(
-                resource=resource,
-                res_iri=res_iri,
-                link_props=link_props,
-                res_iri_lookup=res_iri_lookup,
-                con=con,
-                verbose=verbose
-            )
-        except BaseError as err:
-            print(f'BaseError while updating an XML text of resource "{resource.id}": {err.message}')
-            continue
-        except Exception as exception:
-            print(f'Exception while updating an XML text of resource "{resource.id}": {exception}')
-            continue
+        res_iri = id2iri_mapping[resource.id]
+        existing_resource = con.get(path=f'/v2/resources/{quote_plus(res_iri)}')
+        for link_prop, hash_to_value in link_props.items():
+            existing_values = existing_resource[link_prop.name]
+            if not isinstance(existing_values, list):
+                existing_values = [existing_values, ]
+            for existing_value in existing_values:
+                old_xmltext = existing_value.get("knora-api:textValueAsXml")
+                if not old_xmltext:
+                    continue
+
+                # strip all xml tags from the old xmltext, so that the pure text itself remains
+                pure_text = re.sub(r'(<\?xml.+>\s*)?<text>\s*(.+)\s*<\/text>', r'\2', old_xmltext)
+
+                # if the pure text is a hash, the replacement must be made
+                if pure_text not in hash_to_value:
+                    continue
+                new_xmltext = hash_to_value[pure_text]
```

Comment:
Where does this hash come from?

Reply:
From `stash_circular_references()`, where all XML texts with a salsah-link that leads to circularity are replaced by their hash, and the original text is stashed. Now, the inverse is done: All XML texts that were stashed are taken out, and put to the place where they came from. For this, they are identified by their hash.